### ISSUE TITLE
Large compile times due to explosion of data pathways

### ISSUE THREAD

**Issue by pranv** [Theano/Theano#3396]
Theano constructs a Symbolic expressions of final outputs by following it's evolution in the graph. The number of paths from input variables to loss functions increases if your model uses multiple pathways from input to output, eg: Fork or Split. For something like the inception module, 4 seperate paths exist per module. This results in 4^9 pathways in the case of complete GoogLeNet. This causes massive compile times. Similar explosion of number of pathways occurs, even if the model contains just a few splits in the final layers. 

Constructing these expressions from bottom up fashion, rather than in a top down way should solve the problem.

**Comment by nouiz** [comment]
We don't optimize each path individually. Theano optimization time is
dependent of the number of nodes in the graph, not of the number of path.

What make you think it is the number of path and not the number of nodes
that slow down compilation?

If you have a stand alone GoogleLeNet model, it would help us speed it up
if you make it available. If not, can you update Theano (we speed up
optimization recently, this happen frequently), and run one example with
those theano flags: profile=True,profile_optimizer=True and give us the
result? You don't need to have the function execute, just that they get
compiled. It will tell us which part of the compilation pipeline is the
bottleneck for that case.

Different model trigger the optimizer in different way. So we need that
from different model to optimize it correctly.

**Comment by pranv** [comment]
When compiling the expression, we repeatedly call the output of previous layer. This causes long compile times. 

@nouiz I will add the details you asked shortly

**Comment by benanne** [comment]
When you say 'we' are you talking about Keras specifically? And with 'call' do you mean you call a function that builds the output expression, or are you just reusing the same output expression multiple times?

If the former is the case, have a look at https://github.com/Lasagne/Lasagne/issues/104 and https://github.com/Lasagne/Lasagne/issues/97 . Lasagne originally did something similar and we figured out it's a bad idea (especially when you add things like dropout). You cannot count on Theano to merge all of these common subexpressions as this is error-prone and it takes a lot of time. It's better to just create the output expression of each layer once, and then re-use it.

**Comment by pranv** [comment]
@benanne, **we** referred to keras and **call** to the recursive output function. The lasagne issues and your fixes were the kind of the things I was looking for. Thanks :)

I guess this doesn't concern theano now, since once graph is optimized, this problem is eliminated.

**Comment by nouiz** [comment]
I'm still interrested to see the output of such case with those flags:

profile=True,profile_optimizer=True

just to get an idea of where the time is spent.

By the way, there is a e new PR that should make optimization faster with
cudnn. It  speed up a case from 190s to 12s! But don't expect that speed up
to be general, I think it got a worst case:

https://github.com/Theano/Theano/pull/3403

**Comment by pranv** [comment]
I'll put it up soon

**Comment by nouiz** [comment]
the PR is merged
